compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java

```diff
@@ -2462,6 +2464,100 @@ public void visit(BLangSimpleVariable varNode) {
         varNode.annAttachments.forEach(annotationAttachment -> analyzeNode(annotationAttachment, env));
     }
 
+    private boolean isValidInferredArray(BLangNode node, List<BLangExpression> indexSizes) {
+        switch (node.getKind()) {
+            case INTERSECTION_TYPE_NODE:
+            case UNION_TYPE_NODE:
+                return isValidInferredArray(node.parent, indexSizes);
+            case VARIABLE:
+                BLangSimpleVariable varNode = (BLangSimpleVariable) node;
+                BLangExpression expr = varNode.expr;
+                return validateContext(node.parent) && expr != null && expr.getKind() == NodeKind.LIST_CONSTRUCTOR_EXPR
+                        && validateInitializer(indexSizes, indexSizes.size() - 2, (BLangListConstructorExpr) expr);
+            default:
+                return false;
+        }
+    }
+
+    private boolean validateContext(BLangNode node) {
```

Comment:
Can we choose a better name for this method. If I try to answer the question, Is this function validating something?; current names does not match.

Reply:
Change the method name to `checkContext`. is it okay?

Reply:
Even `checkContext` doesn't reflect the actual check, does it? Just by looking at the name (and even the params) it is not clear what we are checking.

IMO, we should use something like `isValidContextForInferredArray`.